Library/Homebrew/cmd/tests.rb

```diff
@@ -36,18 +24,55 @@ def tests
         args << "TESTOPTS=--name=test_#{test_method}" if test_method
       end
       args += ARGV.named.select { |v| v[/^TEST(OPTS)?=/] }
-      system "bundle", "exec", "rake", "test", *args
 
-      Homebrew.failed = !$?.success?
+      begin
+        sanitize_env!
```

Comment:
This should be placed before the `begin`. Better yet, make `sanitize_env!` accept a block and move the cleanup into that method. As it is responsible for sanitizing the environment, it should be equally responsible for tearing down things afterwards. To be more specific, here's a sketch:

```ruby
def with_sanitized_env
  old_env = ENV.to_hash
  # Current contents of `sanitize_env!`, but use local variables instead of instance variables.

  yield old_home, env_home
ensure
  ENV.replace(old_env)
  # Remaining cleanup code that wipes `env_home`.
end
```

And then use it here like:

```ruby
with_sanitized_env do |old_home, env_home|
  # The gist of the code currently between `begin` and `ensure`.
end
```

(I think this makes for a cleaner control flow.)